### PR TITLE
Wedderburn

### PR DESCRIPTION
**Pull request by piotrmizerka**
@kalmarek I implemented a matrix example as described in my notes: https://github.com/piotrmizerka/postdoc-warsaw-daily-notes/blob/main/pdf/2022_07_28_matrix_Wedderburn.pdf . Now, I think again that my reasoning is correct there (I described the definition of the action of S_3 in line 115 in "experiment_wedderburn.jl". A more detailed explanation below:

I think one has to distinguish between two situations

(1) two actions on R^n and R^m defined by reps rho: G-->GL_n(R), phi: H-->GL_m(R) define the action of GxH on the tensored space R^n\otimes R^m\cong R^{nm} given by (g,h)\mapsto rho(g)\otimes phi(h). I think this is the action you have in mind (in your e-mail from 13 August) in your comment that rho_0 is the representation of S_2x\Sigma.

(2) suppose we have two actions of the same group G, given by reps rho: G-->GL_n(R), phi: H-->GL_m(R). Then we can define the action of G on R^n\otimes R^m\cong R^{nm} by (g,h)\mapsto rho(g)\otimes phi(g).

One can therefore think of two actions: one broader, of GxG on R^{nm} (special case of (1)) and the second (the one we desire I suppose, the case (2)) of G on R^{nm}. 

In your message, you also wrote that one should define the action of \Sigma on {1,2} by pi_3 - but this action would be faithful, whereas ours is not...

Summing up: I think for our purpose (G=<s_1,...,s_n|...>) one shall consider the basis i\otimes e, where i=1,...,n and e\in E (E - e.g. half basis) and define the actions of a subgroup G of S_n by permutations on {1,...,n} and E respectively. These actions define the action on our basis {i\otimes e} (note that the basis is of size n|E|).  So in the example I provided we took G=\Sigma=S_3 acting on E={x,y,z} by permutations, and on {1,2} as describe in line 115. Typically, the set E will be much bigger than the order of permutations we consider.

### CODE REVIEW

**Comment by kalmarek** [comment]
this is what we came up today:
`/src/alphabet_permutation.jl`
```julia
struct AlphabetPermutation{GEl,I} <: SymbolicWedderburn.ByPermutations
    perms::Dict{GEl,Perm{I}}
end
function AlphabetPermutation(
    A::Alphabet,
    Γ::PermutationGroups.AbstractPermutationGroup,
    op,
)
    return AlphabetPermutation(
        Dict(γ => inv(Perm([A[op(l, γ)] for l in A])) for γ in Γ),
    )
end
function AlphabetPermutation(A::Alphabet, W::Constructions.WreathProduct, op)
    return AlphabetPermutation(
        Dict(
            w => inv(Perm([A[op(op(l, w.n), w.p)] for l in A])) for
            w in W
        ),
    )
end

function Base.:^(
    w::Groups.AbstractWord,
    p::PermutationGroups.AbstractPerm,
)
    return typeof(w)([l^p for l in w])
end
function SymbolicWedderburn.action(
    act::AlphabetPermutation,
    γ::GroupElement,
    g::Groups.AbstractFPGroupElement,
)
    G = parent(g)
    w = word(g)^(act.perms[γ])
    return G(w)
end
```

defined somewhere locally:
```
function _conj(
    t::Groups.Transvection,
    σ::PermutationGroups.AbstractPerm,
)
    return Groups.Transvection(t.id, t.i^σ, t.j^σ, t.inv)
end

function _conj(
    t::Groups.Transvection,
    x::Groups.Constructions.WreathProductElem,
)
    tσ = _conj(t, x.perm)
    dual_id = ifelse(t.id == :ϱ, :λ, :ϱ)
    dual_inv = ifelse(t.inv, false, true)
    new_id = isone(x.elts[t.i]) ? t.id : dual_id
    new_inv = isone(x.elts[t.j]) ? t.inv : dual_inv

    return Groups.Transvection(new_id, tσ.i, tσ.j, new_inv)
end

n = 3
G = SAut(FreeGroup(n))
Σ = WreathProduct(SymmetricGroup(2), SymmetricGroup(n))
act = AlphabetPermutation(alphabet(G), Σ, _conj)
```

File: src/wedderburn.jl [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+using SymbolicWedderburn
+using PermutationGroups
+using Groups
+
+struct MatrixAction <: SymbolicWedderburn.ByPermutations end
+struct TensorSupportElement
+    row_generator::Groups.GroupElement
+    column_generator::Groups.GroupElement
+    entry::Groups.GroupElement
+end
```
Comment (kalmarek) [inline src/wedderburn.jl]: we could write this as
```julia
struct TensorSupportElement{B<:AbstractBasis}
    i::Int
    j::Int
    g::Int
    basis::B
end
```

File: src/wedderburn.jl [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    g::Groups.GroupElement
+)
+    SAutFn = parent(g)
+    result = one(SAutFn)
+    word = g.word
+    for i in 1:length(word)
+        result = result*act_on_transvection(σ, SAutFn(word[i:i]))
+    end
+    return result
+end
```
Comment (kalmarek) [inline src/wedderburn.jl]: replaced by `SymbolicWedderburn.action(act, σ, g)`

File: src/wedderburn.jl [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    ::MatrixAction,
+    g::Groups.Constructions.WreathProductElement,
+    tensor_support_element::TensorSupportElement,
+)
+    return TensorSupportElement(
+        wreath_conjugation(g, tensor_support_element.row_generator),
+        wreath_conjugation(g, tensor_support_element.column_generator),
+        wreath_conjugation(g, tensor_support_element.entry)
+    )
+end
```
Comment (kalmarek) [inline src/wedderburn.jl]:
```julia
function SymbolicWedderburn.action(
    act::AlphabetPermutation,
    g::Groups.GroupElement
    tse::TensorSupportElement,
)
    i = SymbolicWedderburn.action(act, tse.i, g)
    j = SymbolicWedderburn.action(act, tse.j, g)
    k = SymbolicWedderburn.action(act, tse.g, g)
    return TensorSupportElement(i, j, k)
end
```

File: src/wedderburn.jl [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+            if i == j
+                for e in half_basis
+                    push!(psd_basis, TensorSupportElement(S[i], S[j], e)) # half_basis on diagonal
+                end
+            end
+        end
+    end
+
+    return constraints_basis, psd_basis
+end
```
Comment (kalmarek) [inline src/wedderburn.jl]:
```julia
function matrix_bases(basis, half_basis, S)
     constr_basis = [TensorSupportElement(s, t, g) for s in S for t in S for g in basis]    
     psd_basis = [TensorSupportElement(s, s, g) for s in S for g in half_basis]
    return constr_basis, psd_basis
end
```

File: src/wedderburn.jl [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+function wedderburn_decomposition_matrix(
+    Σ,
```
Comment (kalmarek) [inline src/wedderburn.jl]:
```suggestion
    action,
    Σ,
```

File: src/wedderburn.jl [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    basis,
+    half_basis,
+    S # stands for the generating set we choose
+)
+    action = MatrixAction()
```
Comment (kalmarek) [inline src/wedderburn.jl]: define externally as `action = AlphabetPermutation(A, Σ, _conj)`

File: src/wedderburn.jl (hunk 1 above)
Comment (kalmarek) [reply src/wedderburn.jl]:
```julia
struct TensorSupportElement{GEl}
     i::GEl
     j::GEl
     k::GEl
end
```
gives much clearer code

File: src/wedderburn.jl [hunk 7]
```diff
@@ -76,3 +76,16 @@ basis, sizes = Groups.wlmetric_ball(S, radius = 2*half_radius)
...
+# first right then left (the classical way) or the other way round.
+S[1]
+S[2]
+elt = S[1]*S[2]
+s1, s2 = Groups.domain(S[1])
+evaluate(elt)
+evaluate(elt) == (s1*s2, s2*s1*s2) # first right then left
+evaluate(elt) == (s1*s2*s1, s2*s1) # first left then right
+# It turns out that the convention is the classical one: compose automorphisms first
+# from the right then from the left.
```
Comment (kalmarek) [inline src/wedderburn.jl]: @piotrmizerka 
```julia
julia> saut = SpecialAutomorphismGroup(FreeGroup(3))
automorphism group of free group on 3 generators

julia> S = gens(saut)
12-element Vector{Automorphism{FreeGroup{Symbol, KnuthBendix.LenLex{Symbol}}, …}}:
 ϱ₁.₂
 ϱ₁.₃
 ϱ₂.₁
 ϱ₂.₃
 ϱ₃.₁
 ϱ₃.₂
 λ₁.₂
 λ₁.₃
 λ₂.₁
 λ₂.₃
 λ₃.₁
 λ₃.₂

julia> g = S[1]*S[2]
 ┌ ϱ₁.₂*ϱ₁.₃:
 │   f1 ↦ f1*f2*f3
 │   f2 ↦ f2
 └   f3 ↦ f3
```
so `g` acts on `domain(g) = (f1, f2, f3)` as follows
 * we first applied `S[1]=ϱ₁.₂` to `(f1, f2, f3)` obtaining `(f1·f2, f2, f3)`
 * we then applied `S[2]=ϱ₁.₃` to `((f1·f2), f2, f3)` obtaining `((f1·f2)·f3, f2, f3)`
as far as I understand this is action **on the right**. Am I mistaken?

File: src/positive_approx_symmetrized.jl [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    end
+
+    cnstrs = constraints(A.mstructure)
+    @assert length(cnstrs) == length(basis(A))
+
+    constraints_basis = w_dec_matrix.basis
+    considered_constraints = Dict{eltype(constraints_basis), Bool}()
+    for constraint in constraints_basis
+        considered_constraints[constraint] = false
+    end
```
Comment (kalmarek) [inline src/positive_approx_symmetrized.jl]: constraints_basis = SymbolicWedderburn.basis(w_dec_matrix)
```suggestion
    constraints_basis = SymbolicWedderburn.basis(w_dec_matrix)
    considered_constraints = Dict(c=>false for c in constraints_basis)
```
Why do we need to store all of these in a dict?

File: src/positive_approx_symmetrized.jl [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    constraints_basis = w_dec_matrix.basis
+    considered_constraints = Dict{eltype(constraints_basis), Bool}()
+    for constraint in constraints_basis
+        considered_constraints[constraint] = false
+    end 
+
+    action = AlphabetPermutation(alphabet(parent(first(S))), Σ, _conj)
+    A_gs = Dict(g => A_g for (A_g, g) in zip(cnstrs, basis(A)))
+
+    for constraint in constraints_basis
```
Comment (kalmarek) [inline src/positive_approx_symmetrized.jl]: shouldn't this loop go over the invariant vectors of `w_dec_matrix`?

File: src/positive_approx_symmetrized.jl [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        if considered_constraints[constraint] == false
+            i, j = word(constraint.row_generator)[1], word(constraint.column_generator)[1]
+            lhs = M[i,j](constraint.entry)
+            δ_i_j_g_invariant = invariant_constraint_matrix(constraint, Σ, action, psd_basis, A_gs, length(S))
+            rhs = 0
+            for (π, ds) in pairs(SymbolicWedderburn.direct_summands(w_dec_matrix))
+                Uπ = SymbolicWedderburn.image_basis(ds)
+                rhs += dot(degree(ds)*Uπ*δ_i_j_g_invariant*Uπ', P[π])
```
Comment (kalmarek) [inline src/positive_approx_symmetrized.jl]:
```suggestion
                rhs += degree(ds)*dot(Uπ*δ_i_j_g_invariant*Uπ', P[π])
```
algebra is associative, but storage isn't ;)

File: src/positive_approx_symmetrized.jl [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+    action = AlphabetPermutation(alphabet(parent(first(S))), Σ, _conj)
+    A_gs = Dict(g => A_g for (A_g, g) in zip(cnstrs, basis(A)))
+
+    for constraint in constraints_basis
+        if considered_constraints[constraint] == false
+            i, j = word(constraint.row_generator)[1], word(constraint.column_generator)[1]
+            lhs = M[i,j](constraint.entry)
+            δ_i_j_g_invariant = invariant_constraint_matrix(constraint, Σ, action, psd_basis, A_gs, length(S))
+            rhs = 0
```
Comment (kalmarek) [inline src/positive_approx_symmetrized.jl]: what is the type of `rhs`?

**Review by kalmarek — CHANGES_REQUESTED** [review]

File: src/wedderburn.jl (hunk 5 above)
Comment (piotrmizerka) [reply src/wedderburn.jl]: we don't need this function to be defined separately

File: src/positive_approx_symmetrized.jl (hunk 11 above)
Comment (piotrmizerka) [reply src/positive_approx_symmetrized.jl]: Instead of `0` one could write `zero(typeof(lhs))`, I guess...

File: src/positive_approx_symmetrized.jl (hunk 9 above)
Comment (piotrmizerka) [reply src/positive_approx_symmetrized.jl]: Yeah. That would be more reasonable... As for now, the version with the iteration over the whole constraint basis is left. I think that the total number of iterations would be O(constraint_basis size)). I guess it is hard to beat this when iterating over all invariant vectors as each invariant vector has length = constraint_basis size and one has to iterate over it somehow...

File: src/wedderburn.jl (hunk 7 above)
Comment (piotrmizerka) [reply src/wedderburn.jl]: I think, I understand the composition differently here... My comprehension of how to apply `S[2]=\rho_1,3` to `f1*f2`: 

`f1*f2` goes to `S[2](f1)*S[2](f2)=f1*f3*f2`.

Hence, assuming action on the right, `(S[1]*S[2])=(f1*f3*f2, f2, f3)`.

On the other hand, if we assume action on the left, then:

1. we apply first `S[2]`: `S[2](f1, f2, f3) = (f1*f3, f2, f3)`,
2. and then we apply `S[1]`: `(S[1]*S[2])(f1, f2, f3) = S[1](S[2](f1, f2, f3)) = S[1](f1*f3, f2, f3) = (f1*f2*f3, f2, f3)`.

File: src/wedderburn.jl [hunk 12]
```diff
@@ -0,0 +1,76 @@
+struct AlphabetPermutation{GEl,I} <: SymbolicWedderburn.ByPermutations
+    perms::Dict{GEl,PermutationGroups.Perm{I}}
+end
+
+function AlphabetPermutation(A::Alphabet, G::Groups.Constructions.WreathProduct, op)
+    return AlphabetPermutation(
+        Dict(
+            g => PermutationGroups.Perm([A[op(l, g)] for l in A.letters]) for
```
Comment (kalmarek) [inline src/wedderburn.jl]: `A.letters` → `A`

File: scripts/SAut_F_4.jl [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+                    g_inv = true
+                    break
+                end
+            end
+            break    
+        end
+    end
+
+    return Groups.Transvection(g_id, g_i, g_j, g_inv)
+end
```
Comment (kalmarek) [inline scripts/SAut_F_4.jl]: `Alphabet` defines a bijection between 1:n and letters of alphabet, ie.
```
A = Alphabet(...)
l = A[1] # → the first letter of A
A[l] # the index of l, i.e. 1
```
Actually it is also possible to do `A[-1]` to obtain __the inverse__ of the first letter (if it exists).

```julia
function determine_transvection(g::Automorphism)
    @assert parent(g) isa Automorphism{<:FreeGroup}
    @assert length(word(g)) = 1
    A = alphabet(parent(g))
    return A[first(word(g))]
end

**Comment by piotrmizerka** [comment]
Coming back to the case when we symmetrize SAutFn with the wreath product. Suppose SAutFn has N generators. Still, I think we have to work out Fox derivatives with F_2N instead of F_N. In the Lyndon's paper, https://www.jstor.org/stable/1969440#metadata_info_tab_contents, he works strictly with the presentation G=F/R, where F is a free group. I think, this setup is quite strict. If we would work with F_N instead of F_2N, we would not have such a situation. For example, suppose x is a generator of SAutFn which we first treat as a generator of F_2N. the inverse of x in SAutFn corresponds to another generator, say y of F_2N, of whom we do know nothing a priori (unless we introduce a specific relation, which we do in a moment). Indeed, the relation we introduce is x*y=1, its Fox derivative over x, d(x*y)/dx equals 1. On the other hand, if we pass right away to F_N, then y is just the inverse of x in F_N. Therefore, in F_N, d(x*y)/dx = d(1)/dx = 0.

File: scripts/SAut_F_4.jl (hunk 13 above)
Comment (kalmarek) [reply scripts/SAut_F_4.jl]: hmm? why resolved when the old `determine_transvection` is still there?

**Comment by kalmarek** [comment]
Hmm... so let's do a concrete example:
* `SAut(F₄)` is generated (as a group!) by 24 elements (12 `ϱ`s and 12 `λ`s)
* (per Lyndon Lemma 5.1) to obtain `²M¹` we need to take `∂Rᵢ/∂sⱼ` for each relation `Rᵢ` defining `SAut(F₄)` and each of 24 generators `sⱼ` of it.

So at least in principle we don't need to take derivatives wrt. the inverses of generators. But somehow we do in here: https://github.com/piotrmizerka/LowCohomologySOS/blob/28ab152fac8e950b107f0475ad8914d4618ff525/scripts/SAut_F_4.jl#L67-L77

Why was it necessary? Per Lyndon paper we don't also need to take derivatives of the relations that are present in the free group, so the example you gave seems not applicable? hmm...

File: scripts/SAut_F_4.jl (hunk 13 above)
Comment (piotrmizerka) [reply scripts/SAut_F_4.jl]: sorry, forgot to change it, now shall be ok ;)

**Comment by piotrmizerka** [comment]
> 

Yes, we don't need to take inverses in order to compute cohomology from the matrices Mi. However, we need inverses for Wedderburn: we want to act on M_n(RG) by the wreath product (which flips the generators among others). That's why we need to add the inverses to the generating set. Now, as we have the symmetrized generating set, according to Lyndon, we work with the free group generated by this symmetrized set - this is F_48. So, in order to obtain SAutF_4 as a quotient, we have to introduce several new relations, e.g. those which tell that the gens corresponding to inverses are the indeed inverses.

**Comment by kalmarek** [comment]
Ok, I understand. 

We probably need to asses if using the wreath product action actually leads to smaller problem (we start with something larger, so we get more symmetries) than using just the symmetric group (we start with something smaller, so we get smaller symmetries).
Usually there's not much sense in first symmetrizing the problem only to use symmetry to reduce its size ;)

But maybe here we gain something more?

File: src/positive_approx_symmetrized.jl [hunk 14]
```diff
@@ -1,55 +1,38 @@
...
 )
-    result = zeros(eltype(first(A_gs)[2]), psd_basis_length, psd_basis_length)
-    for σ ∈ Σ
-        σ_orbit_representative = SymbolicWedderburn.action(action, σ, orbit_representative)
-        σi, σj = word(σ_orbit_representative.i)[1], word(σ_orbit_representative.j)[1]
-        δ_σi_σj = KroneckerDelta{generators_number}(σi, σj)
-        A_σg = A_gs[σ_orbit_representative.k]
-        result += δ_σi_σj⊗A_σg
+    bs = length(A_gs)
+    n = floor(Int, sqrt(div(length(v_inv), bs)))
```
Comment (kalmarek) [inline src/positive_approx_symmetrized.jl]: `isqrt`

File: src/positive_approx_symmetrized.jl [hunk 15]
```diff
@@ -72,78 +55,33 @@ function sos_problem_symmetrized(
...
-
-    direct_summands_pairs = pairs(SymbolicWedderburn.direct_summands(w_dec_matrix))
-    psd_basis_length = length(psd_basis)
-
-    for constraint in w_dec_matrix.basis
-        
-        # if it%100 === 0
-        #     @info it
+    m = size(first(cnstrs))[1] # this is equal to the half_basis' size
+    n = floor(Int, sqrt(div(length(basis(w_dec_matrix)), length(basis(A)))))
```
Comment (kalmarek) [inline src/positive_approx_symmetrized.jl]: `isqrt`

**Review by kalmarek — COMMENTED** [review]
the symmetrized `sos_problem` looks good!

File: .github/workflows/CI.yml [hunk 16]
```diff
@@ -9,7 +9,7 @@ jobs:
       fail-fast: false
       matrix:
         version:
-          - '1.6'
+          - '1.8'
```
Comment (kalmarek) [inline .github/workflows/CI.yml]: why are we no longer testing on julia `1.6`?

File: src/positive_approx_symmetrized.jl [hunk 17]
```diff
@@ -1,5 +1,5 @@
 basis(A::StarAlgebras.StarAlgebra) = A.basis
-basis(w::WedderburnDecomposition) = w.basis
+basis(w::SymbolicWedderburn.WedderburnDecomposition) = w.basis
```
Comment (kalmarek) [inline src/positive_approx_symmetrized.jl]: why not just importing `StarAlgebras.basis` ?

File: src/group_rings.jl [hunk 18]
```diff
@@ -5,6 +5,12 @@ Base.copy(X::AlgebraElement) =
     AlgebraElement(copy(StarAlgebras.coeffs(X)), parent(X))
 
 StarAlgebras.star(g::Groups.AbstractFPGroupElement) = inv(g)
+function StarAlgebras.Basis{I}(basis::AbstractVector) where {I}
+    length(basis) <= typemax(I) ||
+        throw("index type $I is to small for basis of length $(length(basis))")
+    # @assert !(eltype(basis) <: Integer)
+    return StarAlgebras.Basis(basis, Dict(b => I(idx) for (idx, b) in pairs(basis)))
+end
```
Comment (kalmarek) [inline src/group_rings.jl]: hey, hey, hey @piotrmizerka :D the assertion is there because it's needed, not because of a whim :D

`Basis` is a vector which also implements a bijection between indices of `basis` vector and its elements. You can't distinguish between the two if both are integers: if `basis = [2,1]` then is `basis[1]` the first element or the index of `1` in `basis`?

File: src/wedderburn.jl [hunk 19]
```diff
@@ -51,9 +51,9 @@ function subset_permutation(
     g::Groups.GroupElement,
     act::AlphabetPermutation
 )
-    subset_idies = Dict(subset[i] => i for i in Int32.(eachindex(subset)))
+    subset_idies = Dict(subset[i] => i for i in UInt32.(eachindex(subset)))
```
Comment (kalmarek) [inline src/wedderburn.jl]: `UInt32.(eachindex(subset))` materializes the whole vector which you definitely don't want to...

**Comment by kalmarek** [comment]
@piotrmizerka use `git rebase` to move this branch over the master!

**Comment by piotrmizerka** [comment]
> @piotrmizerka use `git rebase` to move this branch over the master!

which branch: Wedderburn or the one for our arxiv article?

**Comment by kalmarek** [comment]
@piotrmizerka move this branch on top of master/main
